src/clj/fluree/db/virtual_graph/bm25/index.clj

```diff
@@ -182,23 +181,48 @@
               adds
               removes))))
 
+(defn upsert-queries
+  [db parsed-query affected-iris]
+  (let [results-ch (async/chan)
+        iri-var    (-> parsed-query :select :subj)]
+    (async/go
+      (loop [[next-iri & r] affected-iris]
+        (if next-iri
+          (let [values  [{iri-var (where/match-iri next-iri)}]
+                pq      (assoc parsed-query :values values)
+                result  (async/<! (exec/query db nil pq))
+                result* (if (vector? result)
+                          (first result)
+                          result)]
+            ;; note that query-result could be a single item if query used :selectOne, or a vector if not
+            (async/>! results-ch (if (nil? result*)
+                                   [::bm25.update/retract {"@id" next-iri}]
+                                   [::bm25.update/upsert result*]))
+            (recur r))
+          (async/close! results-ch))))
+    results-ch))
+
 (defn bm25-upsert
   [bm25 db add removes]
   (let [prop-deps      (property-dependencies bm25)
         affected-sids  (affected-subjs prop-deps add removes)
         affected-iris  (map #(iri/decode-sid db %) affected-sids)
+        item-count     (count affected-iris)
         pq             (parsed-query bm25)
-        iri-var        (-> pq :select :subj)
-        iri-values     (map #(hash-map iri-var (where/match-iri %)) affected-iris)
-        pq*            (assoc pq :values iri-values)
-        upsert-docs-ch (exec/query db nil pq*)]
 
-    (bm25-upsert* bm25 db upsert-docs-ch)))
+        ;; TODO - change below single query to upsert-queries to handle multiple queries
```

Comment:
it looks like this todo no longer applies. there is also commented code on line 214